validator_worker/src/sentry_interface.rs

```diff
@@ -1,143 +1,261 @@
+
+        // validate that we are to validate the channel
+        match channel
+            .spec
+            .validators
+            .into_iter()
+            .find(|&v| v.id == whoami)
+        {
+            Some(v) => {
+                let validator_url = format!("{}/channel/{}", v.url, channel.id);
+
+                Ok(Self {
+                    adapter,
+                    validator_url,
+                    client,
+                    logging,
+                    channel: channel.to_owned(),
+                    config: config.to_owned(),
+                })
+            }
+            None => Err(ValidatorWorkerError::InvalidValidatorEntry(
+                "we can not find validator entry for whoami".to_string(),
+            )),
         }
     }
 
-    pub fn propagate() {
+    pub fn propagate(&self, messages: Vec<MessageTypes>) {
+        let serialised_messages: Vec<String> = messages
+            .into_iter()
+            .map(|message| serde_json::to_string(&message).unwrap())
+            .collect();
 
+        for validator in self.channel.spec.validators.into_iter() {
+            let auth_token = self
+                .adapter
+                .get_auth(&validator)
+                .expect("Failed to get user auth token");
+            match propagate_to(
+                &auth_token,
+                self.config.propagation_timeout,
+                &validator,
+                &serialised_messages,
+            ) {
+                Ok(_) => return,
+                Err(e) => handle_http_error(e, &validator.url),
+            }
+        }
     }
 
-    async pub fn get_latest_msg(from: &str, type: &str) -> impl Future<Output = Result<Vec<Channel>, Error>>  {
+    pub async fn get_latest_msg(
+        &self,
+        from: String,
+        message_type: String,
+    ) -> Result<ValidatorMessageResponse, reqwest::Error> {
         let future = self
             .client
-            .get(format!("{}/validator-messages/{}/{}?limit=1", self.sentry_url, from, type))
+            .get(&format!(
+                "{}/validator-messages/{}/{}?limit=1",
+                self.validator_url, from, message_type
+            ))
             .send()
-            .and_then(|mut res: Response| res.json::<ChannelAllResponse>());
+            .and_then(|mut res: Response| res.json::<ValidatorMessageResponse>())
+            .compat();
 
-        await!(future.compat())
+        await!(future)
     }
 
-    async pub fn get_our_latest_msg(type: &str) -> impl Future<Output = Result<Vec<Channel>, Error>>  {
-        let whoami = adapter.whoami();
-        self.get_latest_msg(whoami, type)
+    pub async fn get_our_latest_msg(
+        &self,
+        message_type: String,
+    ) -> Result<ValidatorMessageResponse, reqwest::Error> {
+        let whoami = self.adapter.whoami();
+        await!(self.get_latest_msg(whoami, message_type))
+    }
+
+    pub async fn get_last_approved(&self) -> Result<LastApprovedResponse, reqwest::Error> {
```

Comment:
Since I am working branched out of this branch, I found out that this is incorrect as `LastApprovedResponse` has a `Vec` of messages, where as this method returns single message. I will probably fix it, but I am noting it for reference.

Reply:
I don't think that's correct

Reply:
I think ApproveState returns Vec<ApproveState> and NewState returns a single message

Reply:
Ok, so I am apparently missing something?
https://github.com/AdExNetwork/adex-validator/blob/master/routes/channel.js#L167

Reply:
this is also an opened question:
https://github.com/AdExNetwork/adex-validator/blob/master/routes/channel.js#L190